### PR TITLE
Improvements for accurate display of 'Welcome View' 

### PR DESCRIPTION
**Pull request by williamsyang-work**
Adds `traceViewer.serverOn` VSCode context.

Replaces the use of `ReactPlaceholderWidget` with a welcome view.

Improves tracking of `trace-explorer.noExperiments` VSCode Context.

image

Signed-off-by: Will Yang <william.yang@ericsson.com>

### CODE REVIEW

File: vscode-trace-extension/src/test/extension-test.spec.ts [hunk 1]
```diff
@@ -16,8 +16,8 @@ test('Open Trace from Explorer', async ({ page }) => {
 
 test('Open Trace from Trace Viewer', async ({ page }) => {
     await page.getByRole('tab', { name: 'Trace Viewer' }).locator('a').click();
-    await page.getByRole('button', { name: 'Opened Traces Section' }).hover();
-    await page.getByRole('button', { name: 'Open Trace' }).click();
+    await page.getByRole('button', { name: 'Welcome' }).hover();
+    await page.getByRole('button', { name: 'Open Folder' }).click();
```
Comment (bhufmann) [inline vscode-trace-extension/src/test/extension-test.spec.ts:20]: I think `Open Trace` is till correct and should not be changed.

File: vscode-trace-extension/package.json [hunk 2]
```diff
@@ -124,43 +124,56 @@
...
         }
       ]
     },
     "viewsWelcome": [
       {
-        "view": "traces",
-        "contents": "No traces found [learn more](https://www.eclipse.org/tracecompass/).\n[Add traces](command:nodeDependencies.addEntry)"
+        "view": "welcome",
+        "name": "Start Server",
+        "contents": "The server is offline. \n[Resume Trace Extension](command:vscode-trace-server.start-if-stopped)",
```
Comment (bhufmann) [inline vscode-trace-extension/package.json:163]: unfortunately, the `vscode-trace-extension` needs to be independent to the `vscode-trace-server` extension. There are users that don't use the `vscode-trace-server` extension and have their own way of starting the trace server. I suggest to look into the external API for starting a trace server. Right now a third-party extension (e.g. `vscode-trace-server`) can register their implementation to start the server. The current API requires a path to be passed. Maybe you can augment the API to add a default start api without passing the path. This should be optional.

File: vscode-trace-extension/package.json [hunk 3]
```diff
@@ -124,43 +124,56 @@
...
+        "name": "Open Trace",
+        "contents": "There are currently no opened traces. \n[Open Folder](command:openedTraces.openTraceFolder)",
+        "when": "trace-explorer.noExperiments || !traceViewer.serverOn"
       },
       {
-        "view": "item properties",
-        "contents": "No properties to show"
+        "view": "welcome",
+        "name": "Documentation",
+        "contents": "To learn more about how to use the Trace Viewer Extension [read our documentation](https://github.com/eclipse-cdt-cloud/theia-trace-extension).",
```
Comment (bhufmann) [inline vscode-trace-extension/package.json:175]: This is great. Suggestion for a subsequent PR. I wonder if there is a need to make the link configurable so that adopters of the viewer can plug-in a link to their documentation.

File: vscode-trace-extension/package.json (hunk 2 above)
Comment (bhufmann) [inline vscode-trace-extension/package.json:163]: `Trace Extension` is not the name that is used for the extension. The published name is `Trace Viewer` and it should be aligned with it: `Resume Trace Viewer`

File: vscode-trace-extension/package.json [hunk 4]
```diff
@@ -124,43 +124,56 @@
...
+        "name": "Start Server",
+        "contents": "The server is offline. \n[Resume Trace Extension](command:vscode-trace-server.start-if-stopped)",
+        "when": "!traceViewer.serverOn"
       },
       {
-        "view": "views",
-        "contents": "No trace selected [learn more](https://www.eclipse.org/tracecompass/)"
+        "view": "welcome",
+        "name": "Open Trace",
+        "contents": "There are currently no opened traces. \n[Open Folder](command:openedTraces.openTraceFolder)",
```
Comment (bhufmann) [inline vscode-trace-extension/package.json:169]: `Open Trace`

File: vscode-trace-extension/package.json (hunk 3 above)
Comment (bhufmann) [inline vscode-trace-extension/package.json:175]: Also remove `Extension` from the contents line: ... use the Trace Viewer [read documentation] ...

File: vscode-trace-extension/package.json (hunk 2 above)
Comment (bhufmann) [inline vscode-trace-extension/package.json:163]: `The server is offline` should be `The trace server is offline`

File: vscode-trace-extension/package.json (hunk 3 above)
Comment (bhufmann) [inline vscode-trace-extension/package.json:175]: The link should point to the `vscode-trace-extension` repository

**Review by bhufmann — CHANGES_REQUESTED** [review]
Very nice improvement. Thank you very much. I have some comments though.

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts [hunk 5]
```diff
@@ -41,7 +41,27 @@ export const addTspClientChangeListener = (listenerFunction: (tspClient: TspClie
 export async function isUp(): Promise<boolean> {
     const health = await getTspClient().checkHealth();
     const status = health.getModel()?.status;
-    return health.isOk() && status === 'UP';
+    const serverIsUp = health.isOk() && status === 'UP';
+    vscode.commands.executeCommand('setContext', 'traceViewer.serverOn', serverIsUp);
+    if (serverIsUp) {
+        // If the server is up, check for opened experiments
+        areExperimentsOpened();
```
Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:48]: This, along with line 58 below in `areExperimentsOpened()`, which calls-back `isUp()`, seem to introduce a potential for an infinite loop, when the server is detected as being "up".

Comment (williamsyang-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:48]: addressed

File: vscode-trace-extension/package.json (hunk 3 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json:175]: addressed

File: vscode-trace-extension/package.json (hunk 2 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: addressed

File: vscode-trace-extension/package.json (hunk 3 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: addressed

File: vscode-trace-extension/package.json (hunk 4 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: addressed

File: vscode-trace-extension/package.json (hunk 2 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: addressed

File: vscode-trace-extension/src/test/extension-test.spec.ts (hunk 1 above)
Comment (williamsyang-work) [reply vscode-trace-extension/src/test/extension-test.spec.ts]: addressed

File: vscode-trace-extension/package.json (hunk 2 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: Looking at the `trace-server-manager` code I see a way to de-couple the starting of a trace server with opening a trace.  However, I will implement this into a separate PR.

For now, I'm simply going to remove the Start Server welcome view from this PR.  The code for this `viewsWelcome` contribution will be simple enough to re-introduce in the future.

**Comment by williamsyang-work** [comment]
There is currently a bug with the 'Open Trace' screen getting stuck after opening a trace in the local use-case.

image

File: vscode-trace-extension/package.json (hunk 3 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: After the `.vsix` file is built for an extension, it can be unzipped and the original package.json file can be modified in post.  This can be done using `jq` or running a node script from the shell.  I have done this to modify the default setting for server path with the vscode-trace-server extension.

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts [hunk 6]
```diff
@@ -41,7 +41,23 @@ export const addTspClientChangeListener = (listenerFunction: (tspClient: TspClie
 export async function isUp(): Promise<boolean> {
```
Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:41]: Since this is exported, and it's getting used more now, I think a more descriptive name would be good, like `isServerUp()` or `isTraceServerUp()`. Specially since the module name (`backend-tsp-client-provider`) does not clarify much, what exactly "is up".

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts [hunk 7]
```diff
@@ -41,7 +41,23 @@ export const addTspClientChangeListener = (listenerFunction: (tspClient: TspClie
...
     const health = await getTspClient().checkHealth();
     const status = health.getModel()?.status;
-    return health.isOk() && status === 'UP';
+    const serverIsUp = health.isOk() && status === 'UP';
+    vscode.commands.executeCommand('setContext', 'traceViewer.serverOn', serverIsUp);
```
Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:45]: Update the JSDoc to mention that the modified context?

Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:45]: For consistency with `TSP`, should we use `traceViewer.serverUp`?

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts [hunk 8]
```diff
@@ -41,7 +41,23 @@ export const addTspClientChangeListener = (listenerFunction: (tspClient: TspClie
...
+    return serverIsUp;
+}
+
+/**
+ * Checks for opened traces and updates the vscode context `trace-explorer.noExperiments`
+ * @returns noExperiments as boolean
+ */
+export async function areExperimentsOpened(): Promise<boolean> {
```
Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:53]: ATM, the returned boolean is not used. I am a bit disturbed by the reversal in meaning of the returned value vs function name. One potential way forward would be for this function to focus on updating the `noExperiment` context and rename it in consequence. e.g. `updateNoExperimentsContext()`

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts (hunk 7 above)
Comment (williamsyang-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts]: addressed

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts (hunk 8 above)
Comment (williamsyang-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts]: Yea this makes sense.  Addressed.

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts (hunk 6 above)
Comment (williamsyang-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts]: addressed

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts (hunk 7 above)
Comment (williamsyang-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts]: addressed

File: vscode-trace-extension/package.json (hunk 3 above)
Comment (bhufmann) [reply vscode-trace-extension/package.json]: Thanks for the suggestion!

File: vscode-trace-extension/package.json (hunk 2 above)
Comment (bhufmann) [reply vscode-trace-extension/package.json]: Thanks. Sounds good to split the PR.

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the updates. The code works well. One thing noticed though, when starting the trace server from command-line, then then there is now way to force refresh the trace explorer. Before it was possible to select the file explorer and then select the trace explorer to do a refresh. Now, this won't cause the trace explorer to  be populated with the opened traces. I think some user might start the server externally and for debugging of the server it would be good to support still that. Could you please look into that?

About the playwright tests, see [here](https://github.com/eclipse-cdt-cloud/vscode-trace-extension?tab=readme-ov-file#running-ui-tests) in the README on how to run it locally and debug.

**Comment by williamsyang-work** [comment]
> Thanks for the updates. The code works well. One thing I noticed though, when starting the trace server from command-line, then then there is now way to force refresh the trace explorer. Before it was possible to select the file explorer and then select the trace explorer to do a refresh. Now, this won't cause the trace explorer to be populated with the opened traces. I think some user might start the server externally and for debugging of the server it would be good to support still that. Could you please look into that?
> 
> About the playwright tests, see [here](https://github.com/eclipse-cdt-cloud/vscode-trace-extension?tab=readme-ov-file#running-ui-tests) in the README on how to run it locally and debug.

I noticed this as well.  This is because with the new `viewsWelcome` there is no functionality to execute a function when the visibility changes.

I also noticed an edge-case where when you start a trace-server with no previously loaded traces, and simultaneously open a trace, the trace will load but the `viewsWelcome` will not go a way.  And, the only way to refresh the state is to open another trace.

I'm going to put this back into a draft for now.  The above case was avoidable with the additional welcome view of `Resume Trace Viewer` functionality.  I'll address your concerns with the ability to refresh the state, as well as this edge case, then I will re-open this PR.

File: vscode-trace-extension/src/test/extension-test.spec.ts [hunk 9]
```diff
@@ -16,7 +16,7 @@ test('Open Trace from Explorer', async ({ page }) => {
 
 test('Open Trace from Trace Viewer', async ({ page }) => {
     await page.getByRole('tab', { name: 'Trace Viewer' }).locator('a').click();
-    await page.getByRole('button', { name: 'Opened Traces Section' }).hover();
+    await page.getByRole('heading', { name: 'TRACE VIEWER: WELCOME' }).hover();
```
[the remark this answers is not in the document]
Comment (bhufmann) [reply vscode-trace-extension/src/test/extension-test.spec.ts:19]: > When I tested it locally I had to revert that change and it worked. What error did you notice that triggered this change?
> 
> Note, when running this test case, the first test case in this file has run before and there is already a trace on the server, and hence the trace explorer opens and not your welcome page.

I had a trace on the server before running the tests. Starting from an empty server, then I see that the welcome view is shown even if the first test case ran before which added a trace.

**Comment by williamsyang-work** [comment]
I added a button that refreshes the state / context of the welcome view.  I could not find a way to refresh the state by navigating away and re-selecting. However, this seems like a common practice for extension development.

Two examples: GitLens and SourceControl:

gitlens-refresh

source-control-refresh

File: vscode-trace-extension/package.json [hunk 10]
```diff
@@ -124,43 +129,50 @@
     },
     "views": {
       "trace-explorer": [
+        {
```
Comment (ngondalia) [inline vscode-trace-extension/package.json:132]: I have some general comments/concerns regarding this approach. I recently raised an issue (https://github.com/eclipse-cdt-cloud/theia-trace-extension/issues/1053) to handle single trace files along with trace folders. With that, i was hoping to achieve two things:

1) Have a separate button to handle trace files (this is simple to achieve with your approach)

2) Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?

Comment (williamsyang-work) [reply vscode-trace-extension/package.json:132]: > I have some general comments/concerns regarding this approach. I recently raised an issue ([eclipse-cdt-cloud/theia-trace-extension#1053](https://github.com/eclipse-cdt-cloud/theia-trace-extension/issues/1053)) to handle single trace files along with trace folders. With that, i was hoping to achieve two things:
> 
> 1. Have a separate button to handle trace files (this is simple to achieve with your approach)
> 2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?

1. I've added this functionality to this PR in a separate commit.  I think there is value in having this available by default, so just put it into the default view.
2. See this comment about modifying the contents of an extension's packate.json post-build: https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/216#discussion_r1520809823.  I'm not sure if this helps you with your use-case, but its currently how I modify contents in order to suit custom needs.

**Comment by williamsyang-work** [comment]
@bhufmann the problem with the UI test has been resolved.  Ready for re-review.

File: vscode-trace-extension/package.json (hunk 10 above)
Comment (bhufmann) [reply vscode-trace-extension/package.json:132]: > 
>     2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?

It's possible. In the packages.json you can have to entries for open directory and open files. Then there can be context flag added in the when-clause to show it or not. 
```json
"viewsWelcome": [
      {
        "view": "welcome",
        "name": "Open Trace Directory",
        "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      },
      {
        "view": "welcome",
        "name": "Open Trace File",
        "contents": "There are currently no opened traces.\n[Open Trace File](command:openedTraces.openTraceFile)",
        "when": "(trace-explorer.noExperiments || !traceViewer.serverUp) && trace-explorer.hasFileOpen"
      },      
      {
        "view": "welcome",
        "name": "Documentation",
        "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      }
    ],
```

This flag can be set programmatically from any extension.
```javascript
vscode.commands.executeCommand('setContext', 'trace-explorer.hasFileOpen', true);
```

@ngondalia  is that what you were looking for?

File: vscode-trace-extension/package.json [hunk 11]
```diff
@@ -169,6 +181,11 @@
           "command": "openedTraces.openTraceFolder",
           "when": "view == traceExplorer.openedTracesView && !trace-explorer.noExperiments",
           "group": "navigation"
+        },
+        {
+          "command": "trace-explorer.refreshContext",
```
[the remark this answers is not in the document]
Comment (williamsyang-work) [reply vscode-trace-extension/package.json:186]: If you look at the last commit on line 211 I added that functionality in.  So, on application startup, there is an initial execution of `trace-explorer.refreshContext`

File: vscode-trace-extension/package.json [hunk 12]
```diff
@@ -124,43 +129,50 @@
...
         }
       ]
     },
     "viewsWelcome": [
       {
-        "view": "traces",
-        "contents": "No traces found [learn more](https://www.eclipse.org/tracecompass/).\n[Add traces](command:nodeDependencies.addEntry)"
+        "view": "welcome",
+        "name": "Open Trace",
+        "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)\n[Open Trace File](command:openedTraces.openTraceFile)",
```
[the remark this answers is not in the document]
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: Sounds good.  It has been removed.

File: vscode-trace-extension/package.json (hunk 10 above)
Comment (ngondalia) [reply vscode-trace-extension/package.json:132]: > > 1. Have a separate button to handle trace files (this is simple to achieve with your approach)
> 
> I've added this functionality to this PR in a separate commit. I think there is value in having this available by default, so just put it into the default view.
> 
> > 2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?
> 
> See this comment about modifying the contents of an extension's packate.json post-build: [#216 (comment)](https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/216#discussion_r1520809823). I'm not sure if this helps you with your use-case, but its currently how I modify contents in order to suit custom needs.

- That wouldn't work for our case becasue we are not publishing our own version of the extension - we will be relying on the marketplace version, so its more preferrable to do it through API/contexts

Comment (ngondalia) [reply vscode-trace-extension/package.json:132]: > > ```
> > 2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?
> > ```
> 
> It's possible. In the packages.json you can have to entries for open directory and open files. Then there can be context flag (`hasFileOpen`) added that is checked in the when-clause. The default value would be `true`.
> 
> ```json
> "viewsWelcome": [
>       {
>         "view": "welcome",
>         "name": "Open Trace Directory",
>         "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)",
>         "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
>       },
>       {
>         "view": "welcome",
>         "name": "Open Trace File",
>         "contents": "There are currently no opened traces.\n[Open Trace File](command:openedTraces.openTraceFile)",
>         "when": "(trace-explorer.noExperiments || !traceViewer.serverUp) && trace-explorer.hasFileOpen"
>       },      
>       {
>         "view": "welcome",
>         "name": "Documentation",
>         "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
>         "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
>       }
>     ],
> ```
> 
> This flag can be set programmatically from any extension.
> 
> ```js
> vscode.commands.executeCommand('setContext', 'trace-explorer.hasFileOpen', false);
> ```
> 
> @ngondalia is that what you were looking for?

- Yeah, that is exactly what i was thinking! The only thing i would suggest is to break it down further to split the two buttons and that can have its own when context clauses:

```
"viewsWelcome": [
      {
        "view": "welcome",
        "name": "Open Trace Directory",
        "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      },
      {
        "view": "welcome",
        "name": "Open Trace File",
        "contents": "There are currently no opened traces.\n[Open Trace File](command:openedTraces.openTraceFile)",
        "when": "(trace-explorer.noExperiments || !traceViewer.serverUp) && trace-explorer.hasFileOpen"
      },      
      {
        "view": "welcome",
        "name": "Documentation",
        "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      }
    ]
```

**Review by bhufmann — DISMISSED** [review]
It looks good to me. Thanks for the UX improvement. 

@ngondalia We'd like to merge it. The `Open Trace File` would have to be added to the welcome page since it's not re-using the PlaceHolderWidget of the traceviewer-react-components. Thanks, Bernd

**Comment by ngondalia** [comment]
> It looks good to me. Thanks for the UX improvement.
> 
> @ngondalia We'd like to merge it. The `Open Trace File` would have to be added to the welcome page since it's not re-using the PlaceHolderWidget of the traceviewer-react-components. Thanks, Bernd

No issues from my end! The "Open Trace File" can be addressed in a separate PR.

File: vscode-trace-extension/package.json [hunk 13]
```diff
@@ -124,43 +129,50 @@
...
       {
-        "view": "views",
-        "contents": "No trace selected [learn more](https://www.eclipse.org/tracecompass/)"
-      },
-      {
-        "view": "item properties",
-        "contents": "No properties to show"
+        "view": "welcome",
+        "name": "Documentation",
+        "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
```
Comment (marcdumais-work) [inline vscode-trace-extension/package.json:174]: I think it would be more appropriate to point to the [extension's README](https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/master/vscode-trace-extension/README.md) vs the repo root

File: vscode-trace-extension/package.json [hunk 14]
```diff
@@ -89,6 +89,11 @@
         "title": "Open Trace",
         "icon": "$(new-folder)"
       },
+      {
+        "command": "trace-explorer.refreshContext",
+        "title": "Refresh",
```
Comment (marcdumais-work) [inline vscode-trace-extension/package.json:94]: image

The name should be more descriptive. Something like "Refresh Trace Server"?

Comment (bhufmann) [reply vscode-trace-extension/package.json:94]: I think it should be `Refresh Trace Viewer` or `Refresh Trace Explorer` because it's the front-end view that is being refreshed using the back-end sever.

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts [hunk 15]
```diff
@@ -36,12 +36,29 @@ export const addTspClientChangeListener = (listenerFunction: (tspClient: TspClie
...
+ * @returns noExperiments as boolean
+ */
+export async function updateNoExperimentsContext(): Promise<boolean> {
+    const response = await getTspClient().fetchExperiments();
+    if (!response.isOk()) {
+        return false;
+    }
+    const noExperiments = !response.getModel()?.length;
+    vscode.commands.executeCommand('setContext', 'trace-explorer.noExperiments', noExperiments);
+    return noExperiments;
```
Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:61]: Return value not used anywhere? Remove if not needed.

File: vscode-trace-extension/package.json (hunk 13 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: addressed

File: vscode-trace-extension/package.json (hunk 14 above)
Comment (williamsyang-work) [reply vscode-trace-extension/package.json]: addressed

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts (hunk 15 above)
Comment (williamsyang-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts]: addressed

File: vscode-trace-extension/package.json (hunk 13 above)
Comment (marcdumais-work) [reply vscode-trace-extension/package.json]: I still see the original URL, pointing to the repo root README (which is more geared towards developers)

File: vscode-trace-extension/src/utils/backend-tsp-client-provider.ts [hunk 16]
```diff
@@ -36,12 +36,29 @@ export const addTspClientChangeListener = (listenerFunction: (tspClient: TspClie
...
     const status = health.getModel()?.status;
-    return health.isOk() && status === 'UP';
+    const serverIsUp = health.isOk() && status === 'UP';
+    vscode.commands.executeCommand('setContext', 'traceViewer.serverUp', serverIsUp);
+    return serverIsUp;
+}
+
+/**
+ * Checks for opened traces and updates the vscode context `trace-explorer.noExperiments`
+ * @returns noExperiments as boolean
```
Comment (marcdumais-work) [inline vscode-trace-extension/src/utils/backend-tsp-client-provider.ts:52]: no longer in-sync

Comment (marcdumais-work) [reply vscode-trace-extension/src/utils/backend-tsp-client-provider.ts]: looks good now

**Review by marcdumais-work — APPROVED** [review]
LGTM, thanks for this nice contribution!

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks for the UX improvement.
